### DISCUSSION TITLE
[Feature] More ways to filter pictures

### DISCUSSION THREAD

**Discussion by RobWW** [damongolding/immich-kiosk#239] [Feature request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

I was wondering if it would be possible to add more ways to filter pictures.
1.  By time:  It would be nice to select only pictures from a specified time period, such as those taken on the present day.  A use case would be to only see pictures from December 25 on Christmas day.  Or limiting pictures to the last several days would be an quick way to select pictures from a recent trip.  In terms of implementation, it would be useful to be able to select by either relative times (last day or 7 days) or specific data ranges (Dec 2 through Dec 5).
2. By string matching in album names:  Instead of selecting each album to include or exclude individually, it would be nice to select groups of albums by matching text in their name.  For example, it would be nice to select all pictures from albums with a title containing 'travel' with only one command.

Thanks again for the great way to display pictures.

**Comment by damongolding** [comment]
How would you envision the date range working? Say if a user has 2 albums and a person set via either the config or url?

Or could this filter act as like a pseudo album? e.g. `album=between_2022-02-02_and_2023-01-23`

I also guess the same could be done with contains `album=contains_travel` though I would be hesitant to allow this "filter" in the url due to url parsing restrictions.

**Comment by RobWW** [comment]
Would a separate criteria, on the same level of albums and people, work for time based selections?  Such as `time=between_2022-02-02_and_2023-01-23` or `time=2024-12-25`?  One advantage of this approach is that it would set a pattern for adding additional criteria later, such as selection by tags if an api becomes available in Immich.

In terms of combining time with other selection criteria, if I am reading at the documentation correctly, album and people selections are currently combined with a logical OR.  In order to maintain consistency, I guess time based selections should be added with an OR as well.  

In practice, I would likely use either album or time based selections, but not both at once.  For me, this feature would be useful when I would like to display a set of pictures, such as from a recent holiday, that I have not yet sorted into albums (an activity that I always seem to be behind on.)  This would be easy to achieve with a URL that cleared the album based selection (my default), and added the time based selection, such as `?album=&time=2024-12-25`

As a side note, can the album selection list be cleared in the current version of kiosk?  Adding `?album=` (that is, album = [empty string]) to the URL does not work for me, it produces an empty selection.  

In terms of url parsing, would restricting the search term to alpha-numeric text work?  This would still greatly speed up album selection.

Thanks!

**Comment by damongolding** [comment]
I'll have to have a think about how it would work in it's final state (`album` vs `time` etc) but I have a prototype working in `damongolding/immich-kiosk-development:0.15.0-alpha.2` that allows `album=between_2022-02-02_and_2023-01-23` if you wanted to have a sneak peek.

Example when `show_album_name` is set to true and using date range 'album'

![image](https://github.com/user-attachments/assets/5f85f479-746e-463b-8914-7f50fb5cdd44)

**Comment by RobWW** [comment]
Thanks, that looks quite useful.

In terms of syntax, I found the meaning of `between` slightly confusing, as using midnight for both the start and end date means the selections includes the start day, but excludes the end day.  (Although I figured this out after a few minutes.)

**Comment by damongolding** [comment]
Nothing is set so we can change the keyword. Could be `from_DATE_to_DATE`?

If you have any thoughts please pass them along 😊

**Comment by RobWW** [comment]
My confusion came from the inconsistency of the first parameter being inclusive, but the second date being exclusive.  I can't think of any concise way to make this difference clear.  `from_DATE_until_DATE` is one idea but does not address the underlying problem.

`after_DATE_before_DATE` makes both days exclusive and is precise, but may trip up people -- dates are not usually given exclusively.  

The other option is to make both dates inclusive.  This would likely work in most cases, but does make selecting only 1 day a bit clunky:  `from_2023-12-25_to_2023_12_25`.  

In the end this might might just be a documentation issue.

**Comment by damongolding** [comment]
Yeah it's how the Immich api works `takenAfter` and `takenBefore` are the option I use. But as I parse the date I can add/remove a day to make it make sense to the user. 

So `date=2022-12-04_to_2022-12-6` would translate to `takenAfter=2022-12-03` `takenBefore: 2022-12-07` if that makes sense. Or something of that kin?

**Reply by RobWW** [reply]
That sounds good to me.  Hopefully it is helpful to others :)
Thanks again!

**Reply by RobWW** [reply]
An update:  I set up a date based album (using the alpha version) over the holidays and it worked nicely.

One variation I used was, on Dec 24, I created an album with a date range from Dec 24 to Dec 31 (ie. at the time, an album with a data range ending in the future).  The effect was that each day over the holidays, as photos were uploaded to the Immich server, the pictures shown by kiosk automatically updated. It was a convenient time saver not needing to sort pictures each day to maintain an album of the holidays.

For most of the year, there is enough cruft in unsorted photos that I would not want an unedited album, but over Christmas I found the time saving of automatically updating greatly outweighed the occasional poor picture (of the floor, etc) being displayed.

**Comment by RobWW** [comment]
To loop back to filtering album names by text strings, I was wondering if a syntax similar that used by either the redirects or multiple users interface would work?  These interfaces essentially let users define text strings, but avoid the problem of parsing arbitrary text in urls.

Something like:

```
filter_expression:
    - name: travel
    - expression: 'travel, trip, vacation'
```

Which could then be called with:
```
http://{URL}?album=all&album_name_filter=travel
```

This concept could also be extended to filter people or tags by text strings.

Besides simplifying configuration, it would make selecting larger groups of albums much less complex.  Currently I have several configurations with lists of 50 plus albums, these large lists essentially require scripts to keep current.

Thanks again!

**Reply by damongolding** [reply]
I do see your predicament. In your mind do the keywords work as a match? e.g. if album name contains X

**Reply by RobWW** [reply]
Yes.  I wish to select a subset of albums. 

Edit:
In the above example, I was intending the syntax to select any album with travel, trip, or vacation in its name.  Although, if multiple expressions are too involved for an initial implementation, being able to filter on just one expression (like vacation) would still be very useful.
```
filter_expression:
    - name: travel
    - expression: 'vacation'
```
